src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

```diff
@@ -420,25 +453,49 @@ private IEnumerable<PackageSource> LoadNuGetSources(PackageSourceLocation packag
             List<Task<(PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages)>> tasks = sources
                 .Select(source =>
                     GetPackageMetadataAsync(source, packageIdentifier, true, linkedCts.Token)).ToList();
-            while (tasks.Any())
-            {
-                Task<(PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages)> finishedTask =
-                    await Task.WhenAny(tasks).ConfigureAwait(false);
-                tasks.Remove(finishedTask);
-                (PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages) result =
-                    await finishedTask.ConfigureAwait(false);
-                if (result.foundPackages == null)
+
+            if (_restoreActionConfig.DisableParallel)
+            {
+                foreach (Task<(PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages)> task in tasks)
                 {
-                    continue;
-                }
+                    var result = task.ConfigureAwait(false).GetAwaiter().GetResult();
+                    if (result.foundPackages == null)
+                    {
+                        continue;
+                    }
 
-                atLeastOneSourceValid = true;
-                IPackageSearchMetadata matchedVersion =
-                    result.foundPackages.FirstOrDefault(package => package.Identity.Version == packageVersion);
-                if (matchedVersion != null)
+                    atLeastOneSourceValid = true;
+                    IPackageSearchMetadata matchedVersion =
+                        result.foundPackages.FirstOrDefault(package => package.Identity.Version == packageVersion);
+                    if (matchedVersion != null)
+                    {
+                        linkedCts.Cancel();
+                        return (result.source, matchedVersion);
+                    }
+                }
+            }
+            else
+            {
+                while (tasks.Any())
                 {
-                    linkedCts.Cancel();
-                    return (result.source, matchedVersion);
+                    Task<(PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages)> finishedTask =
+                        await Task.WhenAny(tasks).ConfigureAwait(false);
+                    tasks.Remove(finishedTask);
+                    (PackageSource source, IEnumerable<IPackageSearchMetadata> foundPackages) result =
+                        await finishedTask.ConfigureAwait(false);
+                    if (result.foundPackages == null)
+                    {
+                        continue;
+                    }
+
+                    atLeastOneSourceValid = true;
```

Comment:
nit: this whole section is duplicated above, can we factor it out to avoid this?